src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj

```diff
@@ -13,56 +14,70 @@
-  </ItemGroup>
+    <PropertyGroup>
+      <_TargetsToRun>Restore;Publish;PublishItemsOutputGroup</_TargetsToRun>
+    </PropertyGroup>
+
+    <MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
+             Properties="RunningPublish=true
+              ;RuntimeIdentifier=$(PackageRID)
+              ;CoreCLRArtifactsPath=$(CoreCLRArtifactsPath)
+              ;R2ROverridePath=$(MSBuildThisFileDirectory)ReadyToRun.targets
+              ;DotNetBuildFromSource=$(DotNetBuildFromSource)"
+             Targets="$(_TargetsToRun)">
+      <Output TaskParameter="TargetOutputs"
+              ItemName="_RawCrossgenPublishFiles" />
+    </MSBuild>
 
-  <Target Name="AddRuntimeFilesToPackage" DependsOnTargets="ResolveRuntimeFilesFromLocalBuild">
     <ItemGroup>
-      <CrossgenFile Include="@(RuntimeFiles)" Condition="'%(Filename)' == 'crossgen'" />
-      <OptimizationDataFile Include="@(RuntimeFiles)" Condition="'%(Filename)' == 'StandardOptimizationData'" />
-      <ReferenceCopyLocalPaths Include="@(RuntimeFiles)" Exclude="@(CrossgenFile);@(OptimizationDataFile)" />
-      <ReferenceCopyLocalPaths TargetPath="tools/" />
+      <_CrossgenPublishFiles Include="@(_RawCrossgenPublishFiles->'%(OutputPath)')"
+                             KeepMetadata="REMOVE_ALL" />
+    </ItemGroup>
+    <ItemGroup Condition="'$(NativeAotSupported)' != 'true'">
+      <FilesToPackage Include="@(_CrossgenPublishFiles)"
+                      Exclude="*.pdb;*.h;*.lib"
+                      TargetPath="tools/" />
+    </ItemGroup>
+    <ItemGroup Condition="'$(NativeAotSupported)' == 'true'">
+      <!-- Treat all native aot assets as native runtime assets -->
+      <FilesToPackage Include="@(_CrossgenPublishFiles->Distinct())"
+                          Condition="'%(Extension)' != '.pdb'"
+                          TargetPath="tools/" />
     </ItemGroup>
   </Target>
 
-  <Target Name="AddFrameworkFilesToPackage" DependsOnTargets="ResolveLibrariesFromLocalBuild">
-    <ItemGroup>
-      <ReferenceCopyLocalPaths Include="@(LibrariesRuntimeFiles)" TargetPath="tools/" />
-    </ItemGroup>
+  <Target Name="RunPublishedCrossgen" AfterTargets="PublishCrossgen"
+          Condition="'$(TargetOs)' == '$(HostOs)' and '$(TargetArchitecture)' == '$(BuildArchitecture)'">
+    <!-- Run the published crossgen if we're not cross-compiling -->
+    <Exec Command="@(FilesToPackage) $(CoreCLRArtifactsPath)IL/System.Private.CoreLib.dll --out $(IntermediateOutputPath)S.P.C.tmp" Condition="'%(FileName)%(Extension)' == 'crossgen2$(ExeSuffix)'"
+          ConsoleToMsBuild="true">
+      <Output TaskParameter="ConsoleOutput" PropertyName="CrossgenOutput" />
+      <Output TaskParameter="ExitCode" PropertyName="CrossgenExitCode" />
+    </Exec>
+    <Error Text="Crossgen failed with code $(CrossgenExitCode), output: $(CrossgenOutput)" Condition="$(CrossgenExitCode) != 0" />
```

Comment:
Why do you need to capture crossgen's output and re-emit it to the console? Doesn't crossgen already do that? When an msbuild Exec task fails, execution is aborted.